lib/template-loader.php

```diff
@@ -176,7 +176,9 @@ function create_auto_draft_for_template_part_block( $block ) {
 				}
 
 				if ( $template_part_file_path ) {
-					$file_contents = file_get_contents( $template_part_file_path );
+					ob_start();
+					include $template_part_file_path;
```

Comment:
Is there a way to "limit" the php functions that are available in the templates?

Reply:
I don't think so 🤔

Reply:
I guess we could "exec php something" and just load the functions we want there but not sure how good is that :P

Reply:
Out of curiosity, why limit the functions that are available?

Reply:
Because we lose the declarativeness of these templates, their results might start to depend on very different things (database...) and the site editor might not reflect what the user would expect. 

I'd rather start very small in what we allow and open later rather than doing the opposite (which is harder).

I think accessing i18n functions and assets are the main identified use-cases for me.

Reply:
Conditions would also be a valid scenario... Example: a "legacy/classic" theme that had 2 layouts switchable using a theme-mod. If that theme wants to migrate to FSE, the only way users of that theme can upgrade without their sites breaking would be if the theme was able to have a condition in its templates
```php
<?php if ( 'boxes' === get_theme_mod( 'posts_layout', 'list' ) : ?>
<!-- Blocks markup -->
<?php else : ?>
<!-- Blocks markup -->
<?php endif; ?>
```

There are dozens of valid scenarios that would require a lot more than just `__()` and `get_template_part`

Reply:
do we want to enable conditions like that? to be honest I'm not sure yet. Feels like themes hacking the system rather than building two themes or two different templates. Questions like that are important to address on a per case scenario rather than giving a free ticket to everything. We're starting from scratch and I'd rather we take this as an opportunity to avoid falling in the current theme state. What make sense for classic themes might not make sense exactly as is for block-based themes.

Reply:
If I'm not mistaken both `exec()` and `eval()` are not allowed in plugins due to security concerns so I'm not sure how good/secure it would be to use `exec` for this... Perhaps we should ping the plugins team on Slack? They've seen lots of things over the years and I'm sure they'd be able to provide some guidance

Reply:
Regardless of what we want the future to look like and the exciting new toys that FSE offers, I think it's important to also remember that there's a decade of themes out there, businesses and people who rely on them. By not allowing some amount of logic, we're basically telling them to trash a decade of work and brand-building for their themes. Businesses and developers who have spent years building a name for their theme **must** be allowed to transition to FSE somehow. If not by allowing logic in the templates, then some other way. But we can't discard the idea of allowing logic in the templates without first having a solid alternative solution & implementation for those who want to bring their themes over to FSE.
What we need is this: `If that setting has value X, load template Y`. If that is somehow implemented, then sure, we can limit this as much as we want.

Reply:
WordPress doesn't have a templating system at all as you might now it from other PHP based frameworks. It's just a set of conventions with a loader. Therefore a WordPress theme template file is in no way different to any other PHP file in WordPress. Meaning you can do anything you want in there.

Since theme files are just PHP scripts, there is no way to keep some PHP functions for running, and others not. Using `eval()` is a no go, because it's a huge security risk. It's widely discouraged in the PHP ecosystem, and [the documentation](https://www.php.net/manual/en/function.eval.php) clearly indicates this. `exec()` is meant for dealing with external programs, so that's not a fit either.

At this point with how FSE is looking, I don't really see a good migration part forward from PHP based themes. Especially because due to the lack of restrictions around what themes can do, themes pretty much do anything they want (see ThemeForest).

But the model for designing a theme with full site editing blocks is just too different. So I'd recommend looking into ways to bring over whatever functionality one might need PHP for, rather than opening Pandora's box. So I agree with @youknowriad on that point.

I also feel that the example that @aristath has posted is hopefully on the way out. I look for to FSE leading us to a place where theme options are longer a thing. Also if you want different displays for a post grid, just create a block for it, and that's it.

Reply:
> I think accessing i18n functions and assets are the main identified use-cases for me.

I agree with this — I think there's value in keeping things simple to start. If it's not possible to limit the PHP that's available here, I'd be more in favor of finding another solution that only allows these two use cases.

Reply:
I agree with the above, and I also hope that the theme-mods things etc are on the way out and soon to be a thing of the past. But hoping is not the same as saying to all theme-developers who have spent a decade that a decade of their work is now useless. 
There needs to be a plan.
If we don't allow logic in templates, then we *must* at least add filters to allow overriding the the template that gets loaded (which HTML file) as well as a filter to actually filter the contents of the template. So basically instead of 
```php
$file_contents = ob_get_clean();
```
in https://github.com/WordPress/gutenberg/pull/25316/files/465a38c1d56973628d0e32d6b5ba17b04446070d#diff-9f399722fe93f7131d23a3944e030b88420d66461c7b776e7b153bafd3ae7100R181
add a filter to make it 
```php
$file_contents = apply_filters( 'the_theme_template_markup', ob_get_clean() );
```
That small change will allow the best of both worlds I think. Theme templates will be pure & clean, and we won't be completely cutting off theme-developers who want to do something more advanced - or even migrate their previous themes to FSE.

Reply:
Actually... If the filter gets added we don't even need to allow translation functions or placeholders for theme-paths etc.
I'd be able to do this in my functions.php 
```php
add_filter( 'the_theme_template_markup', function( $content ) {
  $content = str_replace( '{{POWERED_BY}}', esc_html__( 'Powered by WordPress', 'textdomain' ), $content );
  return $content;
});
```
and that's it... I'd have working translations, placeholders, everything.

Reply:
I do consider that we are talking about two different things here.

The first is the ability to override the template that gets loaded. The current code in Gutenberg isn't very extensible, but it could be made this way. And this would then enable a fallback to whatever PHP template is there. 

So is that what you're looking for? For example if the theme has no `archive.html`, 'archive.php` would be loaded? So what #24834 covers?

The second is the ability to filter the page content, which I guess is less concerned with backwards compatibility, and more with the limitations of the current FSE implementation.

I fear that code snippets like the one above would not only lead to a poor version of a templating language, but also everybody else coming up with their own version of this.

So again I do think it would be better to figure out what we need this for, and find a better way to implement this. Translations are a good point, because without that FSE is a no-go anyway.

Reply:
> I do consider that we are talking about two different things here.

Not necessarily... The way I see it we have 2 options if we want to resolve the translations issue:

1. Implement a custom translation system for templates (something like `{{ __( 'Translate me', 'textdomain' ) }}`)
2. Allow the use of PHP translations in the `.html` files (`<?php esc_html_e( 'Translate me', 'textdomain' ); ?>`

This PR is an implementation of the 2nd method. That implementation coincidentally also solves an issue that - looking ahead - themes will have. Sure, it's not an issue **now**, because there are no FSE themes, but it will be once the big players start migrating their themes to FSE.
[The suggestion above](https://github.com/WordPress/gutenberg/pull/25316#discussion_r513427584) (only allow translation functions) will close the door to the happy side-effect that this PR has. Sure, it's not the intention of this PR to open up the system, but it does. Closing it to only allow `_e`, `esc_html_e` etc and banning all other functions would severely cripple theme-authors. There's absolutely no reason to do that... Yes, there will always be those who write bad code. That's not something we can't avoid. But closing up the implementation to everyone in order to avoid the few people who will not play by the rules is worse.
WordPress _should_ be open, and allowing the system to be modified is more important than keeping the purity of templates that 3rd-party authors write.
The current state of this PR doesn't urge authors to do things they're not "supposed to" do with the vision you or I have in our heads about what the templating system should look like. In fact, most authors won't ever need translations in their templates or even references to images in the theme. 99% of the time, it will be simply copy-paste from the editor. But that 1% when we need something more, we should be able to do it.

Reply:
I don't think there is a need for this. PHP is already allowed in block templates, because blocks are allowed in PHP templates.

```
<?php
get_header();
do_blocks( '<!-- wp:site-title /-->\n<!-- wp:query -->\n<!-- /wp:query -->' );
get_footer();
```

The only real difference is the block templates don't require `<!DOCTYPE html><head><?php wp_head(); ?></head>` and the like. Other than that, it's just two ways to do the same thing and would be confusing for theme developers.

There clearly needs to be a solution for translations and assets but this PR (and option 2 stated [above](https://github.com/WordPress/gutenberg/pull/25316#discussion_r513606594)) isn't it.

Reply:
I've thought a bit more about this, and beyond all the issues relating to placing PHP in FSE templates, there's another downside here.

And that is that if a user edits a template in the Site Editor, they will edit the computation of the PHP code. So whether it's the case of the conditional above, or the translation function, the result of the code will be in the template, and not the code itself.

Related to translations, there are existing solutions like Weglot that rely on DOM parsing to extract strings, and make them translatable. The advantage with Gutenberg is that the content (in block form) is already parsed. So there's no extra step needed to find out what can be translated or not.

So all that is needed is a way for theme authors to mark content as translatable, and there's no need to rely on PHP. What's more is that since Gutenberg knows about the context in which strings are stored, user provided translations can be auto-escaped.

That would relieve theme authors from having to do this manually. Which is a nice change, because plenty of existing templating engines both in PHP and JS have this built-in.

Reply:
> So whether it's the case of the conditional above, or the translation function, the result of the code will be in the template, and not the code itself.

I think that's fine and the way it should be... If the user edits a template it gets stored as "static" content in the db and their edits should persist so no need for anything dynamic

Reply:
> I also hope that the theme-mods things etc are on the way out and soon to be a thing of the past. But hoping is not the same as saying to all theme-developers who have spent a decade that a decade of their work is now useless.
> There needs to be a plan.
> If we don't allow logic in templates, then we must at least add filters to allow […]

Your defense of theme authors is definitely in the right spirit, but I do disagree with "their work is now useless". Traditional themes aren't being phased out, and theme authors and theme users alike can rest assured that what they have now will continue to work and suit their current needs.

The other idea that I need to articulate is this: FSE-ready block-based themes aren't just a technological jump. To use recent tech events as a simile, it's not like moving an ecosystem from x86 to ARM architecture. When switching architectures, compilers will perform different work to generate different app binaries, there will be significant performance benefits, but nothing really changes in how the apps behave, nor will there be changes in their UIs or the APIs they may use. In contrast, traditional WP themes and block-based themes are fundamentally different products as far as the user (not visitors, but site admins) is concerned. We can't just "emulate" traditional themes on the FSE "runtime" like we can emulate x86 programs. The use of conditional layouts and `get_theme_mod` are just excellent illustrations of this: the concepts don't directly map to the user-first site-editing experience that we strive for with FSE.

Thus, we have a duty to keep supporting traditional themes, but it's not our responsibility to automatically port them to the new system. They are different products with different trade-offs. Theme authors interested in porting their themes to blocks will decide, based on the affordances of FSE, what works best for them. In the meantime, we get to fine-tune the best interfaces to solve the _underlying issues_ that e.g. conditional layouts and `get_theme_mod` sought to address in the first place. :)